### PR TITLE
Sum with axis

### PR DESCRIPTION
**Pull request by marcelbischoff**

### CODE REVIEW

File: tinygrad/opsgpu.py [hunk 1]
```diff
@@ -233,13 +233,20 @@ def backward(ctx, grad_output):
...
 
   @staticmethod
   def backward(ctx, grad_output):
-    input, = ctx.saved_tensors
+    input, axis = ctx.saved_tensors
+
+    output = GPUBuffer([1 if axis is None or i in axis else input.shape[i] for i in range(len(input.shape))], hostbuf=grad_output)
+    #make use of broadcasting
+    return binary_op(ctx, 'a+b', output, buffer_zeros(ctx, input.shape))
     ret = buffer_like(ctx, input)
```
Comment (geohot) [inline tinygrad/opsgpu.py:250]: What's all this code after the return?

File: tinygrad/opsgpu.py [hunk 2]
```diff
@@ -233,13 +233,20 @@ def backward(ctx, grad_output):
...
+    if axis is not None:
+        ret.shape = tuple([input.shape[i] for i in range(len(input.shape)) if i not in axis])
+    return ret
 
   @staticmethod
   def backward(ctx, grad_output):
-    input, = ctx.saved_tensors
+    input, axis = ctx.saved_tensors
+
+    output = GPUBuffer([1 if axis is None or i in axis else input.shape[i] for i in range(len(input.shape))], hostbuf=grad_output)
```
Comment (geohot) [inline tinygrad/opsgpu.py:247]: This line is hard to read

File: tinygrad/opsgpu.py [hunk 3]
```diff
@@ -233,13 +233,20 @@ def backward(ctx, grad_output):
...
 class Sum(Function):
   @staticmethod
-  def forward(ctx, input):
-    ctx.save_for_backward(input)
-    return reduce_op(ctx, "out += a", "out", input)
+  def forward(ctx, input, axis=None):
+    ctx.save_for_backward(input, axis)
+    ret = reduce_op(ctx, "out += a", "out", input, axis=axis)
+    if axis is not None:
+        ret.shape = tuple([input.shape[i] for i in range(len(input.shape)) if i not in axis])
```
Comment (geohot) [inline tinygrad/opsgpu.py:240]: spacing

File: tinygrad/opsgpu.py [hunk 4]
```diff
@@ -131,7 +131,7 @@ def reduce_op(ctx, code, code2, inp, axis=None):
     osize = [1]*len(inp.shape)
   else:
     osize = np.array(inp.shape)
```
Comment (geohot) [inline tinygrad/opsgpu.py:133]: np.zeros?

File: tinygrad/opsgpu.py (hunk 2 above)
Comment (marcelbischoff) [reply tinygrad/opsgpu.py]: agree, was surprised that works! I adapt the np.array from reduce

File: tinygrad/opsgpu.py (hunk 1 above)
Comment (marcelbischoff) [reply tinygrad/opsgpu.py]: forgot to collect some garbage

File: tinygrad/opsgpu.py (hunk 4 above)
Comment (marcelbischoff) [reply tinygrad/opsgpu.py:133]: I don't see it

File: tinygrad/opsgpu.py [hunk 5]
```diff
@@ -233,23 +233,25 @@ def backward(ctx, grad_output):
...
-      int gid = get_global_id(0);
-      res_g[gid] = a_g[0];
-    }""")
-    fill(ctx.cl_queue, [np.prod(ret.shape)], None, grad_output.cl, ret.cl)
-    return ret
-
+    input, axis = ctx.saved_tensors
+    
+    if axis is None:
+        bsize = np.ones(len(input.shape))
```
Comment (geohot) [inline tinygrad/opsgpu.py:248]: spacing

File: tinygrad/ops.py [hunk 6]
```diff
@@ -60,14 +60,14 @@ def backward(ctx, grad_output):
...
+  def forward(ctx, input,axis=None):
+    ctx.save_for_backward(input, axis)
+    return np.array([input.sum()]) if axis is None else input.sum(axis=axis)
 
   @staticmethod
   def backward(ctx, grad_output):
-    input, = ctx.saved_tensors
-    return grad_output * np.ones_like(input)
+    input, axis = ctx.saved_tensors
+    return grad_output.reshape([1 if axis is None or i in axis else input.shape[i] for i in range(len(input.shape))]) + np.zeros_like(input)
```
Comment (geohot) [inline tinygrad/ops.py:70]: Hard to read

File: tinygrad/opsgpu.py (hunk 4 above)
Comment (geohot) [reply tinygrad/opsgpu.py:133]: Instead of np.array. You rely on it being 0, so you should be explicit about that.

File: tinygrad/ops.py (hunk 6 above)
Comment (marcelbischoff) [reply tinygrad/ops.py:70]: you prefer to indent this or to use the np solution as osize etc.

Comment (geohot) [reply tinygrad/ops.py:70]: Hmm, I'll merge and refactor 👍

**Comment by geohot** [comment]
Nice!

This is what I changed it to.

https://github.com/geohot/tinygrad/commit/7f249ec76da5b1196aded658c9575e5447aca27c

**Pull request merged**